### PR TITLE
SONiC MAC ACL Yang model update to add support for Source MAC, Destination MAC, Ethertype pattern update, VLAN_ID, PCP, DEI fields.

### PR DESCRIPTION
**Pull request by abhishekd0**
Signed-off-by: Abhishek Dharwadkar <48468617+abhishekd0@users.noreply.github.com>

#### Why I did it
Currently only IP ACL and related model is defined. Support for MAC ACL is missing. Added support for it.

#### How I did it
ACL_RULE table is added with new MAC ACL related fields namely Source MAC, Destination MAC, Ethertype (Pattern updated to match any valid Ethertypes), VLAN, PCP, DEI

#### How to verify it
Yang model tests are attached.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

#### Description for the changelog

Add support for MAC ACL yang model.

### CODE REVIEW

File: src/sonic-yang-models/yang-models/sonic-acl.yang [hunk 1]
```diff
@@ -139,7 +151,7 @@ module sonic-acl {
 
 				leaf ETHER_TYPE {
 					type string {
-						pattern "(0x88CC|0x8100|0x8915|0x0806|0x0800|0x86DD|0x8847)";
+					    pattern "0x[6-9a-fA-F][0-9a-fA-F]{2,3}|0x[1-9a-fA-F][0-9a-fA-F]{3,4}";
```
Comment (bingwang-ms) [inline src/sonic-yang-models/yang-models/sonic-acl.yang:154]: Why the ```ETHER_TYPE``` filed is updated?

Comment (abhishekd0) [reply src/sonic-yang-models/yang-models/sonic-acl.yang:154]: For L2ACL user can specify any valid ethertype for the L2 traffic filtering.

Comment (bingwang-ms) [reply src/sonic-yang-models/yang-models/sonic-acl.yang:154]: I see. Thanks

**Review by bingwang-ms — APPROVED** [review]

**Comment by zhangyanzhao** [comment]
branch conflicts need be resolved before merge.

**Review by qiluo-msft — DISMISSED** [review]

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 [hunk 2]
```diff
@@ -75,8 +75,20 @@ module sonic-acl {
...
 				}
 
-				choice ip_prefix {
-
+				choice src_dst_address {
+					case l2_src_dst_address {
+						when "(/acl:sonic-acl/acl:ACL_TABLE/acl:ACL_TABLE_LIST[ACL_TABLE_NAME=current()/acl:ACL_TABLE_NAME]/acl:type = 'L2')";
+						leaf SRC_MAC {
+							type string {
+								pattern "[0-9a-fA-F]{2}(:[0-9a-fA-F]{2}){5}|[0-9a-fA-F]{2}(:[0-9a-fA-F]{2}){5}/[0-9a-fA-F]{2}(:[0-9a-fA-F]{2}){5}";
```
Comment (qiluo-msft) [inline src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:83]: >pattern [](http://example.com/codeflow?start=8&length=7)

Should we define a shared type, and name it like stypes:mac_and_mask?
I see the pattern repeated below.

**Comment by qiluo-msft** [comment]
@bingwang-ms Could you also help review?

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 2 above)
Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2]: Added common type and used it in acl.yang

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 [hunk 3]
```diff
@@ -144,7 +152,7 @@ module sonic-acl {
 
 				leaf ETHER_TYPE {
 					type string {
-						pattern "(0x88CC|0x8100|0x8915|0x0806|0x0800|0x86DD|0x8847)";
+					    pattern "0x[6-9a-fA-F][0-9a-fA-F]{2,3}|0x[1-9a-fA-F][0-9a-fA-F]{3,4}";
```
Comment (qiluo-msft) [inline src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:155]: >pattern [](http://example.com/codeflow?start=9&length=7)

This change is not related to your PR title or description. And seems the old code is better.
Could you focus on MAC part as original commit?

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 [hunk 4]
```diff
@@ -220,6 +228,26 @@ module sonic-acl {
 				leaf INNER_L4_DST_PORT {
 					type uint16;
 				}
+
+				leaf VLAN_ID {
```
Comment (qiluo-msft) [inline src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:232]: >VLAN_ID [](http://example.com/codeflow?start=9&length=7)

This change is not related to your PR title or description. 
Could you focus on MAC part as original commit?

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 3 above)
Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:155]: This was discussed earlier also. Please see https://github.com/Azure/sonic-buildimage/pull/7917#discussion_r656718927. User should be able to specify any valid Ethertype for L2 ACL.

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 4 above)
Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:232]: VLAN ID is part of the MAC header. This PR adds various other MAC header keys like PCP DEI etc. I will update the description to explicitly list all added fields

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 [hunk 5]
```diff
@@ -220,6 +228,26 @@ module sonic-acl {
...
+					type uint16 {
+						range 1..4093;
```
Comment (qiluo-msft) [inline src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:234]: >4093 [](http://example.com/codeflow?start=15&length=4)

4094 should be good?
How about `0..4095`? the borders are revered but still valid.

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 4 above)
Comment (qiluo-msft) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:232]: Also update the PR title?

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 5 above)
Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2]: Updated to 1-4094 which matches the "config vlan" CLI from [sonic-utilities](https://github.com/Azure/sonic-utilities/blob/1ea88e21f72ccbc73e18d9fa5d1223b378000490/utilities_common/cli.py#L246) .  0 is reserved for priority tag and will be assigned PVID which will be in range 1-4094. 4095 is reserved as well .

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 4 above)
Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:232]: Done

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 3 above)
Comment (ganglyu) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:155]: Is 0x10000 acceptable?
Should we define a shared type?

Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2]: Fixed regex.  This is used only once. Can be moved in future to a shared type if its used/repeated multiple times.

**Review by ganglyu — DISMISSED** [review]

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 3 above)
Comment (ganglyu) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2]: 0x0800 was acceptable, but not now.

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 [hunk 6]
```diff
@@ -144,7 +152,7 @@ module sonic-acl {
 
 				leaf ETHER_TYPE {
 					type string {
-						pattern "(0x88CC|0x8100|0x8915|0x0806|0x0800|0x86DD|0x8847)";
+					    pattern "0x[6-9a-fA-F][0-9a-fA-F]{2}|0x[1-9a-fA-F][0-9a-fA-F]{3}";
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:155]: 0x0800 was acceptable, but not now.

Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2]: Added only those 2 to pattern to maintain backwards compatibility

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 [hunk 7]
```diff
@@ -144,7 +152,7 @@ module sonic-acl {
 
 				leaf ETHER_TYPE {
 					type string {
-						pattern "(0x88CC|0x8100|0x8915|0x0806|0x0800|0x86DD|0x8847)";
+					    pattern "0x[6-9a-fA-F][0-9a-fA-F]{2}|0x[1-9a-fA-F][0-9a-fA-F]{3}|0x0806|0x0800";
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-templates/sonic-acl.yang.j2:155]: 0x0[6-9a-fA-F][0-9a-fA-F]{2}

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/acl.json [hunk 8]
```diff
@@ -657,5 +657,123 @@
...
+    },
+    "ACL_TABLE_L2_ACL_FIELDS": {
+        "sonic-acl:sonic-acl": {
+            "sonic-acl:ACL_RULE": {
+                "ACL_RULE_LIST": [
+                    {
+                        "ACL_TABLE_NAME": "L2ACL",
+                        "SRC_MAC": "00:00:AB:CD:EF:00/FF:FF:FF:00:00:00",
+                        "DST_MAC": "00:00:AB:CD:EF:FF/FF:FF:FF:FF:FF:FF",
+                        "ETHER_TYPE": "0x800",
```
Comment (ganglyu) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/acl.json:669]: Can this field pass the UT?

File: src/sonic-yang-models/yang-templates/sonic-acl.yang.j2 (hunk 7 above)
Comment (abhishekd0) [reply src/sonic-yang-models/yang-templates/sonic-acl.yang.j2]: Leading zeros as no significance and this is represented as hex number. Requester has to take the extra step to add a 0 which is not useful

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/acl.json (hunk 8 above)
Comment (abhishekd0) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/acl.json:669]: updated

**Review by ganglyu — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]
